{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3c4c7f20-668d-4673-b92f-9e95c63e42d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "import tensorflow.compat.v2 as tf\n",
    "import os\n",
    "from collections import defaultdict\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9f7a593b-6a6a-4516-810f-1e36023bcc9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import tensorflow_hub as hub\n",
    "#enformer_model = hub.load(\"https://tfhub.dev/deepmind/enformer/1\").model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b004bd0c-464e-4940-9b77-35d47e995ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEQ_LENGTH = 393216 #Enformer input sequences length\n",
    "N_bins = 896 #Number of Enformer output bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47b4fb7e-1ddd-4740-9d16-4074f628a260",
   "metadata": {},
   "outputs": [],
   "source": [
    "enformer_model_dir = '/lustre/groups/epigenereg01/workspace/projects/vale/tools/enformer/model/'\n",
    "fasta_fa = '/lustre/groups/epigenereg01/workspace/projects/vale/outrider/gene_coords/aml_genes_GRCh37.fa'\n",
    "tss_tsv = '/lustre/groups/epigenereg01/workspace/projects/vale/outrider/gene_coords/transcripts_canonocal_GRCh37.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f6888e2a-9ce6-4a88-80dd-5c4c40d72e86",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants_tsv = '/lustre/groups/epigenereg01/workspace/projects/vale/outrider/promoter_variants/promoters_196K_both_rare_variants.tsv'\n",
    "output_dir = '/lustre/groups/epigenereg01/workspace/projects/vale/outrider/outrider_210223/aml/rare/promoters_2K_both/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0832065-4c50-4d42-9047-1bb78e5e0da0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#targets_idx = np.array(np.arange(0,674)) #DNASE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "72d57fd9-3775-4085-8abe-4290c9bb5405",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants_df = pd.read_csv(variants_tsv, sep='\\t', names=['chrom','pos','ref','alt','vcf_name','geneID','rstart','rstop'])\n",
    "variants_df['pos'] = variants_df['pos']-1 #to 0-based"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cc5f1594-3395-4151-b370-9b4e70483ad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_df = pd.read_csv(tss_tsv, sep=\"\\t\", usecols = [1,3], names=['tss','geneID']) #0-based TSS coordinates for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2e7ec632-144b-4c4b-bc5a-f21660842fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions = variants_df[['geneID','rstart','rstop']].drop_duplicates()\n",
    "regions = regions.merge(tss_df).set_index('geneID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "24b70f97-ac1f-46fa-b7d6-531baffdc1ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_variant(ref, alt, pos, seq, seq_pos):\n",
    "    '''\n",
    "    insert a variant into an existing sequence\n",
    "    seq - array of 'A', 'T', 'C', 'G' or 'N'\n",
    "    seq_pos - absolute positions of sequence bp in the genome\n",
    "    '''\n",
    "    varpos = seq_pos.index(pos) #index inside the sequence of variant position (relative position)\n",
    "    if len(alt)==len(ref):\n",
    "        assert seq[varpos]==ref, 'Wrong reference allele'\n",
    "        seq[varpos] = alt\n",
    "    elif len(alt)>len(ref): #insertion\n",
    "        assert seq[varpos]==ref, 'Wrong reference allele'\n",
    "        seq = seq[:varpos] + list(alt) + seq[varpos+1:]\n",
    "        seq_pos = seq_pos[:varpos] + [seq_pos[varpos]]*len(alt) + seq_pos[varpos+1:] #assign all inserted bases the same position\n",
    "    else: #deletion\n",
    "        assert seq[varpos:varpos+len(ref)]==list(ref), 'Wrong reference allele'\n",
    "        seq = seq[:varpos+1] + seq[varpos+len(ref):]\n",
    "        seq_pos = seq_pos[:varpos+1] + seq_pos[varpos+len(ref):]\n",
    "    return seq, seq_pos\n",
    "\n",
    "def center_around_tss(seq, seq_pos, tss_pos):\n",
    "    '''\n",
    "    center the sequence around the TSS\n",
    "    seq - array of 'A', 'T', 'C', 'G' or 'N'\n",
    "    seq_pos - absolute positions of sequence bp in the genome\n",
    "    tss_pos - absolute position of TSS in the genome\n",
    "    '''\n",
    "\n",
    "    centered_seq = ['N']*SEQ_LENGTH #initialize centered sequence\n",
    "\n",
    "    tss_idx = seq_pos.index(tss_pos) #TSS index in the input sequence\n",
    "\n",
    "    left_seq = seq[max(0,tss_idx-SEQ_LENGTH//2):tss_idx] #part of the input sequence to the left of TSS\n",
    "    right_seq = seq[tss_idx:tss_idx+SEQ_LENGTH//2] #part of the input sequence to the right of TSS\n",
    "    \n",
    "    #insert left and right parts of the input sequence to the centered sequence\n",
    "    centered_seq[SEQ_LENGTH//2:SEQ_LENGTH//2+len(right_seq)] =  right_seq\n",
    "    centered_seq[SEQ_LENGTH//2-len(left_seq):SEQ_LENGTH//2] = left_seq\n",
    "\n",
    "    return centered_seq\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    '''\n",
    "    reverse complement of a given sequence\n",
    "    '''\n",
    "    s = list(map(lambda x:{'A':'T','C':'G','T':'A','G':'C'}.get(x,'N'),seq))\n",
    "    return s[::-1]\n",
    "\n",
    "def roll_seq(seq, shift):\n",
    "    '''\n",
    "    shift a sequence to right (positive shift) or to left (negative shift)\n",
    "    pad with 'N'\n",
    "    '''\n",
    "    if shift>0:\n",
    "        return ['N']*shift + seq[:-shift]\n",
    "    else:\n",
    "        return seq[-shift:] + ['N']*(-shift)\n",
    "    \n",
    "def one_hot(seq):\n",
    "    '''\n",
    "    One-hot encoding in order 'ACGT'\n",
    "    '''\n",
    "    seq = np.array(seq)\n",
    "    s = np.vstack((seq=='A',seq=='C',seq=='G',seq=='T')).astype(int).T\n",
    "    return np.expand_dims(s,0)\n",
    "\n",
    "def enformer_predict(refseq_c, altseq_c):\n",
    "    '''\n",
    "    get enformer predictions for centered reference and alternative sequences\n",
    "    '''\n",
    "    all_pred = []\n",
    "\n",
    "    for seq in refseq_c, reverse_complement(refseq_c), altseq_c, reverse_complement(altseq_c):\n",
    "        for subseq in one_hot(seq), one_hot(roll_seq(seq,47)),one_hot(roll_seq(seq,-47)): \n",
    "            pred = enformer_model.predict_on_batch(subseq)['human'].numpy()\n",
    "            all_pred.append(pred[:,N_bins//2,:]) #only the central bin\n",
    "\n",
    "    all_pred = np.vstack(all_pred)\n",
    "\n",
    "    ref_pred = all_pred[:6,:]#.mean(axis=0) #average for seq, shifted seq (right), shifted seq (left) and reverse complement \n",
    "    alt_pred = all_pred[6:,:]#.mean(axis=0)\n",
    "\n",
    "    #log2fc = np.log2(alt_pred[targets_idx]/ref_pred[targets_idx]).mean()\n",
    "    \n",
    "    return ref_pred, alt_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "82acf60f-45e6-4dfa-9e62-d204dff36916",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "fasta = pysam.FastaFile(fasta_fa)\n",
    "\n",
    "enformer_model = tf.keras.models.load_model(enformer_model_dir).model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a123be37-246e-4364-92b8-e74e512d9424",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene ENSG00000072501 (1/58)\n",
      "53440934 G A\n",
      "53486984 T C\n",
      "53562185 G A\n",
      "53349165 G T\n",
      "53412314 C T\n",
      "53539874 T C\n",
      "53545152 C CT\n",
      "53567490 T C\n",
      "53606804 G A\n",
      "53311956 C CTT\n",
      "53459072 CAGATGGAG C\n",
      "53443264 T A\n",
      "53517305 G A\n",
      "53303632 T TA\n",
      "53336155 T C\n",
      "53530911 C T\n",
      "53263180 G A\n",
      "53273706 CA C\n",
      "53395918 T C\n",
      "53492454 C T\n",
      "53539710 C T\n",
      "53557406 C CT\n",
      "53558336 C T\n",
      "53580141 T A\n",
      "53611215 T C\n",
      "53295449 G A\n",
      "53540339 C T\n",
      "53410926 T A\n",
      "53456700 C T\n",
      "53479215 G C\n",
      "53492657 CAA C\n",
      "53508304 CT C\n",
      "53536734 A ACAAC\n",
      "53555253 AT A\n",
      "53518125 AT A\n",
      "53628846 C CT\n",
      "53337065 TA T\n",
      "53345087 G T\n",
      "53271914 C CT\n",
      "53399282 C CT\n",
      "53449085 G T\n",
      "53388173 G C\n",
      "53341263 T TA\n",
      "53448372 G A\n",
      "53307392 T C\n",
      "53316133 A AG\n",
      "53337331 G A\n",
      "53366891 A G\n",
      "53390785 G A\n",
      "53399303 TG T\n",
      "53419219 G A\n",
      "53428930 T TG\n",
      "53482770 G A\n",
      "53517228 T G\n",
      "53568030 C T\n",
      "53586095 A G\n",
      "53289586 C T\n",
      "53324096 C A\n",
      "53339097 C A\n",
      "53379011 CA C\n",
      "53396400 CA C\n",
      "53407234 T C\n",
      "53455803 C G\n",
      "53476229 C T\n",
      "53518954 A T\n",
      "53563008 G C\n",
      "53277850 T C\n",
      "53394768 G GA\n",
      "53409855 G A\n",
      "53437131 A C\n",
      "53461804 C T\n",
      "53513631 T A\n",
      "53534909 C T\n",
      "53555404 C T\n",
      "53563026 G A\n",
      "53604691 C T\n",
      "53373599 T TTG\n",
      "53408050 G A\n",
      "53472749 C T\n",
      "53477564 A G\n",
      "53494306 CA C\n",
      "53295449 G A\n",
      "53349203 A AG\n",
      "53364523 A G\n",
      "53560006 G T\n",
      "53605490 T G\n",
      "53312349 A G\n",
      "53522252 CA C\n",
      "53277284 G T\n",
      "53285263 G A\n",
      "53289327 C T\n",
      "53335987 G T\n",
      "53347754 A G\n",
      "53377927 G T\n",
      "53382537 G C\n",
      "53420877 A T\n",
      "53421295 CG C\n",
      "53430066 T C\n",
      "53430329 A G\n",
      "53497531 G C\n",
      "53513573 T A\n",
      "53514841 G C\n",
      "53516615 C T\n",
      "53516644 A T\n",
      "53577437 C T\n",
      "53586877 A G\n",
      "53597168 C T\n",
      "53625072 C T\n",
      "53633233 CAAAT C\n",
      "53636746 G A\n",
      "53308194 TAAAAG T\n",
      "53332821 C A\n",
      "53344386 G A\n",
      "53460070 T A\n",
      "53487169 T C\n",
      "53512499 T C\n",
      "53512588 G A\n",
      "53531032 C G\n",
      "53561697 A G\n",
      "53626168 A ATTCT\n",
      "53297073 G C\n",
      "53305566 C T\n",
      "53329099 T A\n",
      "53325383 A AT\n",
      "53325423 T C\n",
      "53364403 T C\n",
      "53365718 T C\n",
      "53375192 GA G\n",
      "53375299 AT A\n",
      "53387605 T C\n",
      "53388173 G C\n",
      "53394593 C T\n",
      "53396286 C G\n",
      "53399990 A G\n",
      "53415916 CA C\n",
      "53417730 A G\n",
      "53420120 A G\n",
      "53481647 A G\n",
      "53487062 T C\n",
      "53536155 TAA T\n",
      "53553924 CA C\n",
      "53562596 T C\n",
      "53606778 G C\n",
      "53609888 C CTCGGCT\n",
      "53261292 C T\n",
      "53265284 A C\n",
      "53388291 G A\n",
      "53388322 T C\n",
      "53542686 T C\n",
      "53627402 C A\n",
      "53627404 A C\n",
      "53283050 G A\n",
      "53562745 T C\n",
      "53262999 C T\n",
      "53483093 C T\n",
      "53508508 G C\n",
      "53565712 TC T\n",
      "53645547 C T\n",
      "53273706 CA C\n",
      "53308207 C T\n",
      "53379471 T C\n",
      "53485248 T C\n",
      "53502219 T TTTTTC\n",
      "53519508 C G\n",
      "53520253 G A\n",
      "53527299 T G\n",
      "53535425 G A\n",
      "53254107 T C\n",
      "53270564 T C\n",
      "53547637 C T\n",
      "53602454 C T\n",
      "53300616 A G\n",
      "53418165 CA C\n",
      "53466359 A G\n",
      "53491395 T C\n",
      "53501644 A AAT\n",
      "53617023 A T\n",
      "53628808 C T\n",
      "53254737 C A\n",
      "53306957 T A\n",
      "53338120 A T\n",
      "53369879 G A\n",
      "53410797 G A\n",
      "53443191 G T\n",
      "53532845 AC A\n",
      "53337331 G A\n",
      "53421867 G C\n",
      "53427876 C T\n",
      "53475283 T A\n",
      "53512845 AC A\n",
      "53533405 A T\n",
      "53537846 T A\n",
      "53581196 T C\n",
      "53605484 G A\n",
      "53625102 G A\n",
      "53639654 A AAACC\n",
      "53309397 G A\n",
      "53355969 G C\n",
      "53366777 ATT A\n",
      "53393526 A G\n",
      "53419376 C T\n",
      "53454117 G T\n",
      "53494727 C T\n",
      "53524175 G C\n",
      "53384173 G T\n",
      "53631298 C T\n",
      "53371840 G A\n",
      "53573209 A C\n",
      "53336156 A G\n",
      "53630032 C T\n",
      "53271169 G A\n",
      "53286786 G A\n",
      "53386098 T C\n",
      "53397418 C T\n",
      "53421162 G A\n",
      "53429861 A C\n",
      "53458245 C T\n",
      "53489837 G GT\n",
      "53498261 G GT\n",
      "53510280 C T\n",
      "53535252 G A\n",
      "53535333 A G\n",
      "53582877 A G\n",
      "53600651 C T\n",
      "53615888 GT G\n",
      "53626219 T C\n",
      "53258976 G A\n",
      "53275281 G C\n",
      "53466301 C T\n",
      "53542116 C T\n",
      "53592520 ATTT A\n",
      "53267265 C T\n",
      "53474977 A C\n",
      "53490556 CT C\n",
      "53496061 TG T\n",
      "53390877 C T\n",
      "53413519 G C\n",
      "53531444 C T\n",
      "53604456 T TCCCCCC\n",
      "53604465 T TCCCCCC\n",
      "53276550 C A\n",
      "53572607 A G\n",
      "53330387 C T\n",
      "53413939 G A\n",
      "53627909 C T\n",
      "53260556 T A\n",
      "53362420 C G\n",
      "53365397 C T\n",
      "53416037 G GT\n",
      "53516144 C T\n",
      "53604626 C T\n",
      "53289586 C T\n",
      "53582853 C T\n",
      "53596834 A G\n",
      "53272951 A T\n",
      "53356362 GA G\n",
      "53442554 G A\n",
      "53456164 A C\n",
      "53513559 T A\n",
      "53553631 C T\n",
      "53585542 T C\n",
      "53625717 C T\n",
      "53295449 G A\n",
      "53337021 T C\n",
      "53408917 G A\n",
      "53416647 C CA\n",
      "53438552 C CA\n",
      "53464214 GA G\n",
      "53553849 C T\n",
      "53565699 C G\n",
      "53568553 C G\n",
      "53369441 A G\n",
      "53377400 C T\n",
      "53490556 CT C\n",
      "53522433 T C\n",
      "53562185 G A\n",
      "53562186 T C\n",
      "53514700 A T\n",
      "53522408 C CA\n",
      "53532642 AT A\n",
      "53584514 C T\n",
      "53303912 G C\n",
      "53312115 C T\n",
      "53337021 T C\n",
      "53408917 G A\n",
      "53416647 C CA\n",
      "53438552 C CA\n",
      "53464214 GA G\n",
      "53565699 C G\n",
      "53568553 C G\n",
      "53258377 G A\n",
      "53272904 C T\n",
      "53297639 G C\n",
      "53300229 A T\n",
      "53554079 C CA\n",
      "53569620 G A\n",
      "53296333 G A\n",
      "53314434 C G\n",
      "53346236 G A\n",
      "53443483 G A\n",
      "53489850 G T\n",
      "53544359 A G\n",
      "53565688 T A\n",
      "53590949 G A\n",
      "53254909 G C\n",
      "53255804 G A\n",
      "53269583 G A\n",
      "53356362 GA G\n",
      "53538370 C T\n",
      "53601499 C G\n",
      "53331388 G A\n",
      "53350562 G C\n",
      "53364523 A G\n",
      "53609128 T C\n",
      "53296075 G GGGGGGAG\n",
      "53358287 A G\n",
      "53423691 T G\n",
      "53433817 C T\n",
      "53470455 A G\n",
      "53487169 T C\n",
      "53497680 A G\n",
      "53278562 A C\n",
      "53416507 G A\n",
      "53481647 A G\n",
      "53542935 G A\n",
      "53631396 T C\n",
      "53299696 C G\n",
      "53342536 G C\n",
      "53351276 C T\n",
      "53356362 GA G\n",
      "53367172 C T\n",
      "53424387 C A\n",
      "53466797 G A\n",
      "53489850 G T\n",
      "53511912 G A\n",
      "53559970 T C\n",
      "53603093 G A\n",
      "53627693 T C\n",
      "53635422 T C\n",
      "53483354 G A\n",
      "53490552 CTTTCTTTCT C\n",
      "53614473 C T\n",
      "53614494 G A\n",
      "53374666 G T\n",
      "53391745 G A\n",
      "53396888 G A\n",
      "53399785 T C\n",
      "53457880 A G\n",
      "53545052 G A\n",
      "53545055 G A\n",
      "53267265 C T\n",
      "53288019 C T\n",
      "53375052 G A\n",
      "53270227 G A\n",
      "53476773 C T\n",
      "53530931 G A\n",
      "53560293 G A\n",
      "53490561 T TTTC\n",
      "53514162 G A\n",
      "53494306 C CA\n",
      "53630292 A G\n",
      "53318043 C T\n",
      "53387913 C G\n",
      "53468855 G A\n",
      "53492656 T TA\n",
      "53492657 C CA\n",
      "53371436 A G\n",
      "53426084 G A\n",
      "53488406 T TG\n",
      "53568243 G C\n",
      "53581030 T C\n",
      "53597120 C A\n",
      "53624273 A G\n",
      "53260092 G A\n",
      "53379313 GGTATTTACCCAAATGAGC G\n",
      "53546523 C T\n",
      "53365507 C CT\n",
      "53322219 G A\n",
      "53407348 C T\n",
      "53423122 C G\n",
      "53489097 C G\n",
      "53535159 C T\n",
      "53569544 T C\n",
      "53619015 G A\n",
      "53367793 C CTTG\n",
      "53401077 G A\n",
      "53580511 T C\n",
      "53602454 C T\n",
      "53633619 A G\n",
      "53624273 A G\n",
      "53336372 G A\n",
      "53337331 G A\n",
      "53475283 T A\n",
      "53512845 AC A\n",
      "53533405 A T\n",
      "53581196 T C\n",
      "53302273 C T\n",
      "53404075 C A\n",
      "53542245 A C\n",
      "53317178 C T\n",
      "53501032 A T\n",
      "53578108 T C\n",
      "53284245 C T\n",
      "53384212 G A\n",
      "53640127 AC A\n",
      "53546726 CAAAAAAA C\n",
      "53589315 G T\n",
      "53306478 C T\n",
      "53315849 G A\n",
      "53416759 TTTC T\n",
      "53416762 C T\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "53416762 is not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-40-d42fa8fdeeda>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     33\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpos\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0malt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 35\u001b[0;31m             \u001b[0maltseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maltseq_pos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minsert_variant\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0malt\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpos\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maltseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maltseq_pos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     36\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m         \u001b[0maltseq_c\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcenter_around_tss\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maltseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maltseq_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mregion_coords\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtss\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# place the alternative sequence in the center\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-16-855a2d23cfb1>\u001b[0m in \u001b[0;36minsert_variant\u001b[0;34m(ref, alt, pos, seq, seq_pos)\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mseq_pos\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mabsolute\u001b[0m \u001b[0mpositions\u001b[0m \u001b[0mof\u001b[0m \u001b[0msequence\u001b[0m \u001b[0mbp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mgenome\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     '''\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0mvarpos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mseq_pos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpos\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#index inside the sequence of variant position (relative position)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0mseq\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mvarpos\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Wrong reference allele'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: 53416762 is not in list"
     ]
    }
   ],
   "source": [
    "N_genes = len(variants_df.geneID.unique())\n",
    "\n",
    "for gene_idx, geneID in enumerate(variants_df.geneID.unique()[52:53]):\n",
    "\n",
    "    print(f'gene {geneID} ({gene_idx+1}/{N_genes})')\n",
    "\n",
    "    refseq = fasta.fetch(reference = geneID).upper() # reference sequence around TSS\n",
    "    \n",
    "    refseq = list(refseq) # string to list\n",
    "\n",
    "    region_coords = regions.loc[geneID] # start/end coordinates of reference sequence and TSS position\n",
    "\n",
    "    refseq_pos = list(range(region_coords.rstart,region_coords.rstop)) # all positions in the reference sequence\n",
    "\n",
    "    refseq_c = center_around_tss(refseq, refseq_pos, region_coords.tss) # place the reference sequence in the center\n",
    "\n",
    "    gene_df = variants_df[variants_df.geneID==geneID] # only variants for given gene\n",
    "    \n",
    "    gene_df = gene_df.drop_duplicates(['vcf_name','chrom','pos']) # exclude overlapping variants (rare)\n",
    "\n",
    "    gene_preds = defaultdict(dict) # predictions for given gene, each entry is a vcf name\n",
    "    \n",
    "    N_samples = len(gene_df.vcf_name.unique())\n",
    "    \n",
    "    for sample_idx, sample in enumerate(gene_df.vcf_name.unique()):\n",
    "        \n",
    "        #print(f'sample {sample} ({sample_idx+1}/{N_samples})')\n",
    "        \n",
    "        altseq, altseq_pos = list(refseq), list(refseq_pos) # initialize altseq with refseq\n",
    "        \n",
    "        for pos,ref,alt in gene_df.loc[gene_df.vcf_name==sample,['pos','ref','alt']].values:\n",
    "            \n",
    "            print(pos,ref,alt) \n",
    "            \n",
    "            altseq, altseq_pos = insert_variant(ref,alt,pos,altseq, altseq_pos)\n",
    "        \n",
    "        altseq_c = center_around_tss(altseq, altseq_pos, region_coords.tss) # place the alternative sequence in the center\n",
    "        \n",
    "        #mutation_at_tss = (gene_df.pos-region_coords.tss==0).sum()\n",
    "        \n",
    "        #if not mutation_at_tss:\n",
    "        #    assert refseq_c[SEQ_LENGTH//2]==altseq_c[SEQ_LENGTH//2], 'ref/alt mismatch at TSS'\n",
    "        \n",
    "\n",
    "        #ref_pred, alt_pred =  enformer_predict(refseq_c, altseq_c)\n",
    "        \n",
    "        #gene_preds[sample]['ref'] = ref_pred\n",
    "        #gene_preds[sample]['alt'] = alt_pred\n",
    "        \n",
    "    #with open(output_dir + geneID + '.pickle', 'wb') as f:\n",
    "        #pickle.dump(gene_preds, f)\n",
    "\n",
    "#enformer_log2fc = get_log2fc(refseq_c, altseq_c)\n",
    "#predictions.append((geneID, sample, enformer_log2fc))\n",
    "        \n",
    "#pd.DataFrame(predictions, columns=['geneID','vcf_name','enformer_log2fc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "acf85aea-0df8-4cc3-ab57-b1ed42df1f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open(output_dir + 'geneID.pickle', 'rb') as f:\n",
    "#        a=pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a0ed0319-1601-42e5-93cc-8b10a18b3bb0",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'builtin_function_or_method' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-39-7159417c0e85>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0maltseq_pos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m53416762\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: 'builtin_function_or_method' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "altseq_pos.index[53416762]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef9df2e9-1300-439a-924f-771b88db00cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
